transport.go

```diff
@@ -96,6 +103,9 @@ type transport struct {
 	serverConfig *quic.Config
 	clientConfig *quic.Config
 	gater        connmgr.ConnectionGater
+
+	holePunchingMx sync.Mutex
+	holePunching   map[peer.ID]map[*net.UDPAddr]context.CancelFunc
```

Comment:
Should this be a `map[*net.UDPAddr]map[peer.ID]context.CancelFunc`? 
f two peers are behind the same NAT (and therefore have the same `net.Addr`), we only need to hole punch once.

Reply:
Actually, depending on the behaviour of the NAT, we might have to hole punch twice because they will have different ports mapped.

Reply:
But the port is part of the `net.UDPAddr`. On the other hand, since two peers behind the same NAT will most likely use different ports, there's not much to gain from deduplicating here.

Reply:
yeah, we can't really have two peers on the same addr, port -- at least not with our code.